### ISSUE TITLE
Improve continuous benchmarking with Bencher

### ISSUE THREAD

**Issue by epompeii** [MystenLabs/fastcrypto#769]
Hey `fastcrypto` team!
I came across [your white paper](https://eprint.iacr.org/2024/442.pdf), and I think you all have built a pretty nice continuous benchmarking site.

I just wanted to reach out because I'm the maintainer of an open source continuous benchmarking tool called Bencher: https://github.com/bencherdev/bencher

It looks like you all currently [only benchmark releases](https://github.com/MystenLabs/fastcrypto/blob/4e43631858e0c3062cd2d2f5bf83fb0428c3b0cf/.github/workflows/benchmarking.yml#L4). Though, I may be missing something.
Bencher would allow you to track your benchmarks over time, compare the performance of pull requests, and catch performance regressions before they get merged.

I would be more than happy to answer any questions that you all may have!

**Comment by jonas-lj** [comment]
Nice!

Yea, we don't run the benchmarks on all PRs because it takes a long time to run them. But it sounds very interesting with better reporting and comparison over time.

I'm curious to hear what your experience is doing benchmarks online as part of CI? For us, performance varies quite a lot, making it a bit difficult to detect small changes in performance.

**Comment by epompeii** [comment]
> Yea, we don't run the benchmarks on all PRs because it takes a long time to run them.

Yeah, this can definitely be a blocker. I think the most common thing I've seen is only running a subset of benchmarks on PRs to at least cover the critical path.

> I'm curious to hear what your experience is doing benchmarks online as part of CI?

There are a few ways to handle this. In order of most to least effective:

1. Use a bare metal runner ($100/month)
2. Use an instruction count based benchmarking harness (in addition to a wall clock based benchmarking harness)
3. Use [statistical continuous benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#statistical-continuous-benchmarking) on shared CI runners
4. Use [relative continuous benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#relative-continuous-benchmarking) on shared CI runners
5. Run a nightly benchmarking job that then does a git bisect to find performance regressions

**Comment by jonas-lj** [comment]
Thanks, that's

> > Yea, we don't run the benchmarks on all PRs because it takes a long time to run them.
> 
> Yeah, this can definitely be a blocker. I think the most common thing I've seen is only running a subset of benchmarks on PRs to at least cover the critical path.
> 
> > I'm curious to hear what your experience is doing benchmarks online as part of CI?
> 
> There are a few ways to handle this. In order of most to least effective:
> 
> 1. Use a bare metal runner ($100+/month)
> 2. Use an instruction count based benchmarking harness (in addition to a wall clock based benchmarking harness)
> 3. Use [statistical continuous benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#statistical-continuous-benchmarking) on shared CI runners
> 4. Use [relative continuous benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#relative-continuous-benchmarking) on shared CI runners
> 5. Run a nightly benchmarking job that then does a git bisect to find performance regressions

Thanks! That's great advise.
